sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UpdateAttributeNullability.scala

```diff
@@ -37,7 +37,7 @@ object UpdateAttributeNullability extends Rule[LogicalPlan] {
     case p if !p.resolved => p
     // Skip leaf node, as it has no child and no need to update nullability.
     case p: LeafNode => p
-    case p: LogicalPlan =>
+    case p: LogicalPlan if p.childrenResolved =>
```

Comment:
This is a concerning change in light of the class comment. If failing to apply this rule can cause wrong results, how sure are we that the rule doesn't need to be applied when children are unresolved?

Reply:
If a child is not resolved, then the next line will fail because the children do not have expression IDs. If the children are never resolved, then the query will fail the resolution check. If the children are eventually resolved, then this simply runs later, when the next line will not fail.

This was a preexisting bug that was uncovered by this PR and is caused by not throwing an analysis exception immediately in `ResolveRelations`. There have been several updates like this to fix places that didn't check child resolution.

Reply:
Ok, that makes sense.

Reply:
why is this needed when the first case is `case p if !p.resolved => p`? Do you mean there is a plan that is resolved but its children are not?

Reply:
@cloud-fan, feel free to look into that, but this case was matching with an unresolved child that did not have expression IDs. I thought the best approach was to require resolved children, but if you'd like to suggest a different approach I can update here.

Otherwise, this solves the problem and it is clearly correct to only access expression IDs when children are resolved.

Reply:
I'm not sure what is the problem this resolved. When we reach here, `p` is already resolved, which usually means `p.childrenResolved` is always true.

Reply:
@cloud-fan, not necessarily. Some plans may not check that children are resolved. `CheckAnalysis` accounts for this case by using `foreachUp`:

```
    plan.foreachUp {
      case o if !o.resolved =>
        failAnalysis(s"unresolved operator ${o.simpleString(SQLConf.get.maxToStringFields)}")
      case _ =>
    }
```